File: Patterns/Manuscript/manuscript/code/java/DesignPatterns/src/main/java/observer/Teacher.java
package observer;

import java.util.ArrayList;
import java.util.List;

class Teacher {
    private List<Student> students = new ArrayList<>();

    public void addStudent(Student student) {
        System.out.println("Teacher added a student " + student.getClass().getName());
        students.add(student);
    }

    public void removeStudent(Student student) {
        System.out.println("Teacher removed a student " + student.getClass().getName());
        students.remove(student);
    }

    public void notifyStudents() {
        System.out.println("Teacher is notifying students.");
        for (Student student : students) {
            student.receiveUpdate();
        }
    }
}

File: code/csharp/DesignPatterns/observer/Teacher.cs
using System.Collections.Generic;

namespace Observer
{
    public class Teacher
    {
        private List<IStudent> _students = new List<IStudent>();

        public void AddStudent(IStudent student)
        {
            System.Console.WriteLine($"Teacher added a student {student.GetType().Name}");
            _students.Add(student);
        }

        public void RemoveStudent(IStudent student)
        {
            System.Console.WriteLine($"Teacher removed a student {student.GetType().Name}");
            _students.Remove(student);
        }

        public void NotifyStudents()
        {
            System.Console.WriteLine("Teacher is notifying students.");
            foreach (var student in _students)
            {
                student.ReceiveUpdate();
            }
        }
    }
}
